<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>resevol: an R package for spatially explicit models of pesticide resistance given evolving pest genomes • resevol</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="resevol: an R package for spatially explicit models of pesticide resistance given evolving pest genomes">
<meta property="og:description" content="resevol">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">resevol</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/resevol.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/advanced_techniques.html">Advanced techniques</a>
    </li>
    <li>
      <a href="../articles/evolutionary_algorithm.html">The evolutionary algorithm</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://bradduthie.github.io/resevol/notebook/" class="external-link">Lab notebook</a>
</li>
<li>
  <a href="https://github.com/bradduthie/resevol" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>resevol: an R package for spatially explicit
models of pesticide resistance given evolving pest genomes</h1>
                        <h4 data-toc-skip class="author">A Bradley
Duthie</h4>
            
            <h4 data-toc-skip class="date">2022-08-06</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bradduthie/resevol/blob/HEAD/vignettes/resevol.Rmd" class="external-link"><code>vignettes/resevol.Rmd</code></a></small>
      <div class="hidden name"><code>resevol.Rmd</code></div>

    </div>

    
    
<hr>
<div class="section level2">
<h2 class="unnumbered" id="abstract">Abstract<a class="anchor" aria-label="anchor" href="#abstract"></a>
</h2>
<ol style="list-style-type: decimal">
<li>The evolution of pesticide resistance is a widespread problem with
potentially severe consequences for global food security.</li>
<li>We introduce the resevol R package, which simulates individual-based
models of pests with evolving genomes that produce complex, polygenic,
and covarying traits affecting pest life history and pesticide
resistance.</li>
<li>Simulations are modelled on a spatially-explicit and highly
customisable landscape in which crop and pesticide application and
rotation can vary, making the package a highly flexible tool for both
general and targetted models of pest management and resistance
evolution.</li>
<li>We present the key features of the resevol package and demonstrate
its use for a simple example simulating pests with two covarying
traits.</li>
<li>The resevol R package is open source under GNU Public License. All
source code and documentation are available on GitHub.</li>
</ol>
<p><strong>Key words:</strong> <em>pest management</em>, <em>food
security</em>, <em>ecological modelling</em>, <em>pesticide</em>,
<em>resistance evolution</em></p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Insect resistance to pesticides is a wicked and widespread problem
<span class="citation">(<a href="#ref-Gould2018" role="doc-biblioref">Gould, Brown, and Kuzma 2018</a>)</span>.
Predicting, identifying, and ultimately delaying the evolution of
insecticide resistance is therefore critical for ensuring sustainable
global food security <span class="citation">(<a href="#ref-Hawkins2019" role="doc-biblioref">Hawkins et al. 2019</a>; <a href="#ref-Lykogianni2021" role="doc-biblioref">Lykogianni et al.
2021</a>)</span>. To manage pesticide resistance, various strategies
have been proposed to disrupt or weaken selection. These strategies
often focus on varying the location and timing of pesticide application,
and therefore varying the strength of selection for pesticide
resistance, so that pest susceptibility to one or more pesticides is
maintained <span class="citation">(<a href="#ref-Ives2002" role="doc-biblioref">Ives and Andow 2002</a>; <a href="#ref-Liang2013" role="doc-biblioref">Liang et al. 2013</a>; <a href="#ref-Carriere2016" role="doc-biblioref">Carrière, Fabrick, and Tabashnik 2016</a>; <a href="#ref-Liang2018" role="doc-biblioref">Liang, Tang, and Cheke
2018</a>; <a href="#ref-South2018" role="doc-biblioref">South and
Hastings 2018</a>; <a href="#ref-Sudo2018" role="doc-biblioref">Sudo et
al. 2018</a>; <a href="#ref-Saikai2021" role="doc-biblioref">Saikai,
Hurley, and Mitchell 2021</a>)</span>. Pesticide resistance management
has mostly been associated with the effect of single genes. In such
cases, resistance alleles have binary effects on phenotype, enabling
resistant phenotypes to arise from genetic changes at single or a small
number of loci <span class="citation">(<a href="#ref-Raymond2001" role="doc-biblioref">Raymond et al. 2001</a>; <a href="#ref-Ffrench-Constant2004" role="doc-biblioref">Richard H.
Ffrench-Constant, Daborn, and Le Goff 2004</a>; <a href="#ref-Ffrench-Constant2013" role="doc-biblioref">Richard H.
Ffrench-Constant 2013</a>)</span>. Nevertheless, resistance attributable
to polygenic effects is also well-established <span class="citation">(<a href="#ref-Via1986" role="doc-biblioref">Via 1986</a>; <a href="#ref-Firko1990" role="doc-biblioref">Firko and Hayes 1990</a>; <a href="#ref-Busi2013" role="doc-biblioref">Busi, Neve, and Powles
2013</a>; <a href="#ref-Haridas2018" role="doc-biblioref">Haridas and
Tenhumberg 2018</a>; <a href="#ref-Green2020" role="doc-biblioref">Green, Stenberg, and Lankinen 2020</a>)</span>. And
the relevance of polygenic resistance is likely to further increase
given a rising interest in biological controls for sustainable crop
protection <span class="citation">(e.g., <a href="#ref-Green2020" role="doc-biblioref">Green, Stenberg, and Lankinen 2020</a>; <a href="#ref-Elnahal2022" role="doc-biblioref">Elnahal et al.
2022</a>)</span>. Developing strategies to maximise the efficacy of
these tools is critical, and such strategies should be well-informed by
predictions made from detailed, quantitative genetic models. Here we
introduce the resevol R package as a tool for building such models and
simulating pest management.</p>
<p>The resevol package applies individual-based modelling and a
quantitative genetics approach to simulate the evolution of a pest
population on a changing landscape. A focal goal of the software is to
model traits with a pre-specified, but potentially evolving, covariance
structure. To achieve this goal, each individual has a genome with <span class="math inline">\(L\)</span> loci that underlie a set of <span class="math inline">\(T\)</span> potentially evolving traits.
Pleiotropic loci can vary in their effects on the direction and
magnitude of polygenic traits, causing population-wide trait covariance
to arise mechanistically from the underlying genetic architecture of
individuals. To achieve this, two separate steps are necessary. First, a
genetic algorithm is used to find a network of internal nodes that map
standard random normal loci values to covarying traits. Values used to
map loci to traits are incorporated into individual genomes. Second, a
population of asexual or sexual individuals is initialised and simulated
on a spatially explicit landscape separated into distinct units (e.g.,
farms). Land units can apply one of up to 10 pesticides, and one of up
to 10 crops at a given time. Pesticides and crops rotate independently
within farms over time in a pre-specified way. The resevol package can
thereby model complex and evolving agricultural pest traits over
realistic landscapes that undergo different pesticide use and crop
regimes.</p>
</div>
<div class="section level2">
<h2 id="mine_gmatrix">Covarying pest quantitative traits<a class="anchor" aria-label="anchor" href="#mine_gmatrix"></a>
</h2>
<p>The first step of simulation is building individual genomes. This
step is separate because it is computationally intensive, and genomes
that are built might need to be inspected and stored. High computation
time is due to the mechanistic nature of how genomes and covarying
traits are modelled. Instead of imposing a trait covariance structure
directly, an evolutionary algorithm is used to find a network that maps
standard random normal values (loci) to covarying values (traits). This
is useful because it allows genomes to model potentially evolving
physiological constraints, and trade-offs among traits, from the bottom
up. Since multiple networks can potentially map loci to the same trait
covariance structure, it is possible to replicate evolution with
different randomly generated genetic architectures. This approach to
modelling individual genomes and traits thereby increases the complexity
of questions that can be addressed for simulating evolution in
agricultural pests.</p>
<div class="figure">
<img src="resevol_files/figure-html/network-1.png" alt="**Figure 1**. *Example network mapping loci (green circles) to traits (red diamonds) through an intermediate set of hidden layers (blue squares) in the `mine_gmatrix` function. Individual genomes in the resevol R package consist of standard random normal values for loci, real values for black arrows linking nodes, and real values for traits. Values shown for loci and arrows are an example for illustration.*" width="576"><p class="caption">
<strong>Figure 1</strong>. <em>Example network mapping loci (green
circles) to traits (red diamonds) through an intermediate set of hidden
layers (blue squares) in the <code>mine_gmatrix</code> function.
Individual genomes in the resevol R package consist of standard random
normal values for loci, real values for black arrows linking nodes, and
real values for traits. Values shown for loci and arrows are an example
for illustration.</em>
</p>
</div>
<p>Conceptually, the relationship between individual genotypes and
traits is defined by a network connecting loci to traits through a set
of hidden internal nodes (Figure 1). Values for loci are randomly drawn
from a standard normal distribution, <span class="math inline">\(\mathcal{N}\left(0, 1\right)\)</span>. Links
between loci, internal nodes, and traits, can take any real value and
are represented by black arrows in Figure 1. Traits can take any real
value, and are calculated as the summed effects of all preceding nodes
(i.e., the blue squares immediately to the left of the traits in Figure
1). Mathematically, loci are represented by a row vector of length <span class="math inline">\(L\)</span>. Effects of loci on the first layer of
internal nodes (black arrows emanating from loci in Figure 1) are
represented by an <span class="math inline">\(L \times T\)</span>
matrix, and transitions between internal nodes, and between the last set
of internal nodes and the final traits, are represented by <span class="math inline">\(T \times T\)</span> matrices. For the example in
Figure 1,</p>
<p><span class="math display">\[
\begin{pmatrix}
  0.1, &amp; -0.1, &amp; 0.2 \\
\end{pmatrix}
\begin{pmatrix}
  0.4, &amp;  0.1 \\
-1.2, &amp; -0.6 \\
  0.2, &amp;  2.2 \\
\end{pmatrix}
\begin{pmatrix}
   1.3, &amp; -0.5 \\
  -0.1  &amp;  0.0 \\
\end{pmatrix}
\begin{pmatrix}
   2.2, &amp;  0.9 \\
   1.6  &amp;  0.8 \\
\end{pmatrix} =
\begin{pmatrix}
  0.2998, &amp; 0.1081 \\
\end{pmatrix}.
\]</span></p>
<p>Values mapping loci to traits become part of an individual’s genome,
so the genome for the individual represented by Figure 1 is stored in
the model as shown below:</p>
<pre><code>0.1, -0.1, -0.2, 0.4, 0.1, -1.2, -0.6, 0.2, 2.2, 1.3, -0.5, -0.1, 0.0, 2.2, 0.9, 1.6, 0.8</code></pre>
<p>Individuals with different loci can therefore have different
covarying traits that are constrained by the network structure encoded
in each genome. Individuals can be haploid (as in Figure 1) or diploid
(in which case, allele values are summed at homologous loci).</p>
<p>An evolutionary algorithm is used to find appropriate values that
produce covarying traits from loci (see Supporting Information 1 for
details). Evolutionary algorithms are heuristic tools that can simulate
adaptive evolution to find solutions for a broad range of problems <span class="citation">(<a href="#ref-Hamblin2013" role="doc-biblioref">Hamblin 2013</a>; <a href="#ref-Duthie2018" role="doc-biblioref">Duthie et al. 2018</a>)</span>. In the resevol
package, the <code>mine_gmatrix</code> function runs an evolutionary
algorithm and requires the argument <code>gmatrix</code>, which
specifies the desired trait covariance matrix. The function initialises
a population of <code>npsize</code> networks (Figure 1), and this
population evolves until some maximum generation (<code>max_gen</code>)
or minimum expected network stress (<code>term_cri</code>) is met. In a
single generation of the algorithm, values crossover between networks
and mutate. Next, trait covariances produced for each network are
estimated by initalising <code>indivs</code> individuals with loci
sampled from a standard normal distribution. Network stress is
calculated as the logged mean squared deviation between estimated
covariances and those in <code>gmatrix</code>. Tournament selection
<span class="citation">(<a href="#ref-Hamblin2013" role="doc-biblioref">Hamblin 2013</a>)</span> is then used to determine
the networks for the next generation of the algorithm. Networks with the
lowest estimated stress have the highest fitness, so these networks are
disproportionately represented in the next generation. Throughout the
evolutionary algorithm, the lowest stress network is saved and returned
upon network termination. The robustness of this network’s stress to
sets of individuals with different loci values can be tested using the
<code>stress_test</code> function.</p>
<p>An example run of <code>mine_gmatrix</code> with the same number of
loci and internal nodes as in Figure 1 is shown below for traits that
have an intended covariance of -0.4:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trait_covs</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="fl">1</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>;
<span class="va">new_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mine_gmatrix.html">mine_gmatrix</a></span><span class="op">(</span>loci <span class="op">=</span> <span class="fl">3</span>, layers <span class="op">=</span> <span class="fl">2</span>, gmatrix <span class="op">=</span> <span class="va">trait_covs</span>, 
                            max_gen <span class="op">=</span> <span class="fl">1000</span>, term_cri <span class="op">=</span> <span class="op">-</span><span class="fl">6.0</span>, prnt_out <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>;</code></pre></div>
<p>The code above found a genome that produced the following expected
trait covariances:</p>
<pre><code><span class="co">##            [,1]       [,2]</span>
<span class="co">## [1,]  1.0008181 -0.3925799</span>
<span class="co">## [2,] -0.3925799  1.0271340</span></code></pre>
<p>The mean deviation between elements of the above matrix and the
identity matrix provided by <code>trait_covs</code> is 2.1176024^{-4}.
Lower values of <code>term_cri</code> and higher values of
<code>max_gen</code> will result in a lower stress, but this will
require more computation time, especially if the number of traits is
high. Similarly, higher values of <code>indivs</code> will result in
more accurate estimations of true stress, but this also requires more
computation time. Additional arguments to <code>mine_gmatrix</code> can
also be used to improve the performance of the evolutionary algorithm
(see Supporting Information 1).</p>
</div>
<div class="section level2">
<h2 id="simulating">Simulating landscape-level pesticide resistance<a class="anchor" aria-label="anchor" href="#simulating"></a>
</h2>
<p>The full output of <code>mine_gmatrix</code> is passed to the
<code>run_farm_sim</code> function, which initialises and simulates an
evolving population of pests on a changing landscape for any natural
number of time steps (<code>time_steps</code>). In this section, we
explain the landscape, pest ecology, and evolving pest traits.</p>
<div class="section level3">
<h3 id="landscape">Landscape<a class="anchor" aria-label="anchor" href="#landscape"></a>
</h3>
<p>Landscapes are spatially explicit and initialised in one of two ways.
First, a landscape can be built from the arguments <code>xdim</code>,
<code>ydim</code>, and <code>farms</code>. These arguments specify the
dimensions of the landscape and the number of farms on it. Contiguous
rectangular farms of roughly equal size are generated on the landscape
using a splitline algorithm. Second, a custom landscape can be input
using the <code>terrain</code> argument, which takes a matrix with
elements that include integers 1 to <code>farms</code>. Each value
defines a unique farm, but values do not need to be contiguous. These
‘farms’ could even model non-farmland (e.g., water, roads), if
pesticides and crops on them are invisible to pests (see Supporting
Information 2). This <code>terrain</code> customisation therefore allows
for a high degree of landscape detail, and offers the potential for
modelling real-world landscapes from raster images. Edge effects are set
using the <code>land_edge</code> argument. Edge options include
<code>"leaky"</code> (pests disappear from the landscape),
<code>"reflect"</code> (pests bounce off of the edge),
<code>"sticky"</code> (pests stick to the edge), or <code>"torus"</code>
(pests that move off of one edge return on the opposite side of the
landscape).</p>
<p>Each farm can hold one pesticide and one crop type in any time step.
The <code>crop_init</code> and <code>pesticide_init</code> arguments
initialise one of <code>crop_number</code> crops and one of
<code>pesticide_number</code> pesticides for each farm, respectively
(maximum of 10 each). Initialisation can be random for each farm with
equal probability, or it can be set using a vector of length
<code>farms</code> in which vector elements define the initialised crop
or pesticide number. After initialisation, crops and pesticides rotate
once ever <code>crop_rotation_time</code> and
<code>pesticide_rotation_time</code> time steps, respectively. Each of
these rotation time arguments can take either an integer value from 1-3,
or a matrix. Integer values specify (1) no rotation, (2) random
transition from one type to another, or (3) cycling through each
available crop or pesticide in numeric order. Square matrices can be
used to define the probability that a given crop or pesticide in row
<span class="math inline">\(i\)</span> transitions to that in column
<span class="math inline">\(j\)</span>. Hence, any possible Markov chain
can be used to model transition between crop or pesticide types on
farms.</p>
</div>
<div class="section level3">
<h3 id="ecology">Pest ecology<a class="anchor" aria-label="anchor" href="#ecology"></a>
</h3>
<p>Individual pests can be modelled to have several reproductive systems
and life histories. Pest reproductive system can be specified using the
<code>repro</code> argument, which accepts <code>"asexual"</code>
(haploid), <code>"sexual"</code> (monoecious), and
<code>"biparental"</code> (dioecious). For <code>"sexual"</code> pests,
the <code>selfing</code> argument specifies if self-fertilisation is
(<code>TRUE</code>) or is not (<code>FALSE</code>) allowed. At the start
of a simulation, pests are initialised in a random location. Initialised
pests are age of zero if <code>rand_age = FALSE</code> or a random age
from zero to <code>max_age</code> if <code>rand_age = TRUE</code>.
Following initialisation, a single time step proceeds with landscape
change (see above), pest aging and metabolism, feeding, pesticide
consumption, movement, reproduction, mortality, and immigration (Figure
2). Feeding, pesticide consumption, movement, and reproduction all
depend on pest age. Pests feed and consume pesticide from ages
<code>min_age_feed</code> to <code>max_age_feed</code>, move from ages
<code>min_age_move</code> to <code>max_age_move</code>, and reproduce
from ages <code>min_age_reproduce</code> to
<code>max_age_reproduce</code> (all inclusive). Food accumulated is lost
during aging if <code>baseline_metabolism &gt; 0</code> and pest age is
within <code>min_age_metabolism</code> and
<code>max_age_metabolism</code>. The option to set minimum and maximum
ages for events makes it possible to model pests with much different
life histories <span class="citation">(e.g., <a href="#ref-Sudo2018" role="doc-biblioref">Sudo et al. 2018</a>)</span>.</p>
<div class="figure">
<img src="resevol_files/figure-html/flowchart-1.png" alt="**Figure 2**. *Overview of simulated events in the resevol R package. Note that metabolism, feeding, pesticide consumption, movement, and reproduction are all subject to a minimum and maximum pest age. Consequently, simulation order might not reflect the order of events from the perspective of a focal pest (e.g., pests might move from ages 1-2, but only feed from ages 2-4). Crops and pesticides are also not necessarily rotated in each time step (see Landscape). Statistics collected within a time step are printed to a CSV file.*" width="768"><p class="caption">
<strong>Figure 2</strong>. <em>Overview of simulated events in the
resevol R package. Note that metabolism, feeding, pesticide consumption,
movement, and reproduction are all subject to a minimum and maximum pest
age. Consequently, simulation order might not reflect the order of
events from the perspective of a focal pest (e.g., pests might move from
ages 1-2, but only feed from ages 2-4). Crops and pesticides are also
not necessarily rotated in each time step (see Landscape). Statistics
collected within a time step are printed to a CSV file.</em>
</p>
</div>
<p>In each time step, pests feed in a random order, consuming the crop
on their landscape cell. Pests consume an amount of the crop as
specified by the <code>food_consume</code> argument, which takes a
vector with as many elements as their are crops (i.e., if
<code>crop_number = 2</code>, then <code>food_consume</code> has two
elements, the first and second defining consumption of crops 1 and 2,
respectively). If crop amount on the landscape cell exceeds pest
consumption ability, then pests consume their maximum amount, and this
amount is removed from the landscape cell. If crop amount is less than
pest consumption ability, then pests consume whatever crop is left and
crop amount is reduced to zero. Pesticide consumption works identically
to crop consumption, except that the amount of pesticide on a landscape
cell is not decreased. Pests simply consume an amount of pesticide as
specified by the <code>pesticide_consume</code> argument, which also
takes a vector with as many elements as their are pesticides. Hence,
each pest can potentially feed and be affected by the pesticide of their
focal landscape cell.</p>
<p>After interacting with their landscape cell, pests can move. Each
pest visits a number of cells during movement, as is specified by the
parameter <code>movement_bouts</code>. Individual movement bouts occur
in a random order across pests. During a movement bout, a pest can
travel to any cell within a value defined by
<code>movement_distance</code> from their current location, which could
include their current location (i.e., moving zero distance). Upon
arrival to a cell, a pest can feed if they are of an appropriate feeding
age and <code>feed_while_moving = TRUE</code>. The pest also consumes
pesticide if they are of the appropriate age and
<code>pesticide_while_moving = TRUE</code>. Having pests feed and
consume pesticide in a random order while moving among landscape cells
can model a population competing for food and encountering pesticides on
a shorter time scale than an individual time step.</p>
<p>After all pests finish moving, pests reproduce. Offspring production
is possible for asexual, monoecious, or female pests. Pest expected
offspring number is defined by a fixed parameter if
<code>reproduction_type = "lambda"</code>, or is calculated from the
amount of food consumed if
<code>reproduction_type = "food_based"</code>. The former requires
specifying <code>lambda_value</code>, which becomes the rate parameter
for sampling offspring number from a Poisson distribution. The latter
requires specifying a real value for <code>food_needed_repr</code>,
which is the amount of food needed to produce one offspring. For
food-based reproduction, the total amount of food consumed is divided by
<code>food_needed_repr</code>, then floored to determine offspring
number. Sexual reproduction requires a mate of reproductive age that is
either monoecious or male, and within range of the reproducing focal
pest (potential including the focal pest, if
<code>selfing = TRUE</code>). A potential mate is within range if it is
within an integer number of cells from the focal pest, as defined by
<code>mating_distance</code> (e.g., if <code>mating_distance = 0</code>,
then mates must share a landscape cell). All available potential mates
sire offspring with equal probability, and reproducing pests are assumed
to mate multiply (i.e., paternity is a fair raffle for all offspring).
If a carrying capacity at birth is set (<code>K_on_birth &gt; 0</code>)
and total offspring number in the population exceeds this capacity, then
offspring are removed at random until they are within carrying capacity.
A real value <code>immigration_rate</code> specifies the rate parameter
for Poisson random sampling of the number of immigrants added to the
population at the end of a time step. Immigrants are initialised in the
same way as pests are at the start of the simulation. Hence, any spatial
structure or evolution that occurs during the simulation does not affect
immigrant locations, genomes, or traits.</p>
</div>
<div class="section level3">
<h3 id="evolution">Pest evolution<a class="anchor" aria-label="anchor" href="#evolution"></a>
</h3>
<p>Pest genomes evolve in a complex and highly mechanistic way.
Offspring inherit genome values from their parent(s) with the
possibility for mutation and recombination; offspring traits are then
calculated from their newly initialised genomes. For asexually and
sexually reproducing pests, genomes are haploid and diploid,
respectively. Asexually reproducing pests receive the full genomes of
their parent, while sexually reproducing pests receive half of their
alleles from each parent. Each diploid parent contributes one half of
their genome, effectively modelling a genome with a single chromosome.
Crossover occurs at each position of the genome with a probability of
<code>crossover_pr</code>, so complete recombination is also possible if
<code>crossover_pr = 0.5</code>. For both haploids and diploids, each
genome value then mutates independently with a probability of
<code>mutation_pr</code>, which can be set to any real number from 0-1.
If a genome value mutates, then a new value is randomly sampled from a
standard normal distribution. If <code>mutation_type = 0</code>, then
this new value replaces the old value, and if
<code>mutation_type = 1</code>, then the new value is added to the old
value. After mutation, genome values are used to calculate trait
values.</p>
<p>Evolution of the genetic architecture linking loci to traits can be
constrained by disabling mutation in genome values linking loci,
internal nodes, and traits (i.e., ‘network values’ represented by arrows
in Figure 1). While mutation at loci (green circles in Figure 1) is
always possible as long as <code>mutation_pr &gt; 0</code>, the number
of intermediary layers for which network values can mutate is
constrained by <code>net_mu_layers</code>. If
<code>net_mu_layers = 0</code>, then no network values can mutate, but
higher integer values cause mutation to occur at network value layers
from loci to traits (if <code>net_mu_dir = 1</code>) or traits to loci
(if <code>net_mu_dir = 0</code>). For example, if
<code>net_mu_layers = 2</code> and <code>net_mu_dir = 1</code>, then the
network values linking loci to the first internal node, and the first
internal node to the second, can mutate (i.e., first two columns of
arrows in 1, but not those linking the second internal node to traits).
This allows pest traits to evolve with varying degrees of constraint on
the covariance between traits. Low <code>net_mu_layers</code> values
model strong genetic constraints, while high values model high
evolvability of trait covariances.</p>
<p>Finally, evolving and covarying traits can be used in place of fixed
parameters described in pest ecology. This is done by substituting
<code>"Tj"</code> as an argument input in place of a numeric value,
where <code>j</code> represents the trait number. For example, the
argument <code>move_distance = "T1"</code> will make Trait 1 the
movement distance for individuals. The argument
<code>food_consume = c("T2", "T3")</code> will set Traits 2 and 3 to
define the amount of food of types 1 and 2 that can be consumed by a
pest, respectively. Up to 10 of the following parameters can be replaced
with evolving traits: <code>move_distance</code>,
<code>food_needed_surv</code>, <code>pesticide_tolerated_surv</code>,
<code>food_needed_repr</code>, <code>pesticide_tolerated_repr</code>,
<code>mating_distance</code>, <code>lambda_value</code>,
<code>movement_bouts</code>, <code>metabolism</code>,
<code>food_consume</code>, and <code>pesticide_consume</code>. Mean
values of traits can be set using the argument <code>trait_means</code>,
which accepts a vector of the same length as the number of evolving
traits such that indices correspond to trait numbers. The resevol
package can thereby simulate agricultural pests with complex and
co-evolving traits, and potentially evolving trait covariances, under a
range of possible pest life histories.</p>
</div>
</div>
<div class="section level2">
<h2 id="output">Simulation output<a class="anchor" aria-label="anchor" href="#output"></a>
</h2>
<p>Simulation output is typically large, so output is printed in two CSV
files, both of which are created in the working directory. The first
file “population_data.csv” prints population level data over time,
including population size, mean age, sex ratio, mean food and pesticide
consumed of each type, mortality rate, and mean trait values. The second
file “individuals.csv” prints all information, including full genomes
and traits (columns), for every individual (rows) in the population. The
printing of individual level data is disabled by default. It can be
turned on for all time steps by setting <code>print_inds = TRUE</code>,
but this should be done with caution because it can create extremely
large files. Instead, individual level data can be printed for only the
final time step by setting <code>print_last = TRUE</code>. Output
produced by <code>run_farm_sim</code> is a list of two elements, which
includes a vector of parameter values used in the simulation and the
final state of the landscape as an array.</p>
</div>
<div class="section level2">
<h2 id="simulation">Example of individual-based simulations<a class="anchor" aria-label="anchor" href="#simulation"></a>
</h2>
<p>In resevol v0.3, the <code>run_farm_sim</code> function includes 68
arguments, which specify a wide range of possible simulation conditions
affecting landscape and pest characteristics. These arguments are
explained the package documentation, and in Supporting Information 2,
which demonstrates an advanced case study with a custom landscape and
complex pest genomes and life history. Here we focus on a simple
simulation with asexually reproducing pests that have three loci and two
traits (Figure 1). We use the the genome generated in
<code>new_network</code> from <a href="#mine_gmatrix">Section 2</a>, in
which traits 1 and 2 have variances of 1.0008181 and 1.027134,
respectively, with a covariance of -0.3925799.</p>
<p>Traits 1 and 2 will define the rate at which pests consume two
separate pesticides, so we model a system in which there is a potential
trade-off for pesticide susceptibility. We use a simple <span class="math inline">\(64 \times 64\)</span> cell landscape with nine
farms. Each farm grows the same crop and uses one of two randomly
intialised pesticides, which are rotated every 16 time steps. Hence, we
can conceptualise 16 time steps as a single growing season.</p>
<p>In our example, each cell produces four crop units, all of which can
potentially be consumed. Pests consume up to one unit of crop on their
landscape cell per time step. Pest survival and reproduction is
food-based, and pests need to consume one unit of crop by age two to
survive and reproduce. Pests initialised at the start of the simulation
are randomly assigned an age from 0-4 with equal probability. Pests have
a maximum age of four, and they feed in ages 0-2, move up to two cells
in ages 3-4, and reproduce in age four. Pests can consume pesticides in
ages 0-2, and if they consume any pesticide, then they will die. In each
time step, a mean of 10 immigrants arrive. We simulate 160 time steps
(10 growing seasons) using the function below:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_farm_sim.html">run_farm_sim</a></span><span class="op">(</span>mine_output <span class="op">=</span> <span class="va">new_network</span>, repro <span class="op">=</span> <span class="st">"asexual"</span>, 
                    pesticide_number <span class="op">=</span> <span class="fl">2</span>, pesticide_init <span class="op">=</span> <span class="st">"random"</span>, 
                    pesticide_consume <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T1"</span>, <span class="st">"T2"</span><span class="op">)</span>, farms <span class="op">=</span> <span class="fl">9</span>,
                    pesticide_rotation_time <span class="op">=</span> <span class="fl">16</span>, pesticide_rotation_type <span class="op">=</span> <span class="fl">3</span>,
                    pesticide_tolerated_surv <span class="op">=</span> <span class="fl">0</span>, pesticide_per_cell <span class="op">=</span> <span class="fl">1</span>,
                    crop_rotation_time <span class="op">=</span> <span class="fl">16</span>, crop_number <span class="op">=</span> <span class="fl">1</span>, crop_per_cell <span class="op">=</span> <span class="fl">4</span>, 
                    food_consume <span class="op">=</span> <span class="fl">1</span>, reproduction_type <span class="op">=</span> <span class="st">"food_based"</span>, 
                    food_needed_surv <span class="op">=</span> <span class="fl">1</span>, food_needed_repr <span class="op">=</span> <span class="fl">1</span>, max_age <span class="op">=</span> <span class="fl">4</span>, 
                    min_age_feed <span class="op">=</span> <span class="fl">0</span>, max_age_feed <span class="op">=</span> <span class="fl">2</span>, min_age_move <span class="op">=</span> <span class="fl">3</span>, 
                    max_age_move <span class="op">=</span> <span class="fl">4</span>, min_age_reproduce <span class="op">=</span> <span class="fl">4</span>, print_gens <span class="op">=</span> <span class="cn">FALSE</span>,
                    max_age_reproduce <span class="op">=</span> <span class="fl">4</span>, age_pesticide_threshold <span class="op">=</span> <span class="fl">2</span>, 
                    rand_age <span class="op">=</span> <span class="cn">TRUE</span>, move_distance <span class="op">=</span> <span class="fl">2</span>, immigration_rate <span class="op">=</span> <span class="fl">10</span>, 
                    time_steps <span class="op">=</span> <span class="fl">160</span>, print_last <span class="op">=</span> <span class="cn">TRUE</span>, xdim <span class="op">=</span> <span class="fl">64</span>, ydim <span class="op">=</span> <span class="fl">64</span>,
                    trait_means <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.1</span><span class="op">)</span>, land_edge <span class="op">=</span> <span class="st">"torus"</span><span class="op">)</span>;</code></pre></div>
<p>Any arguments to <code>run_farm_sim</code> not included above are set
to default values. <a href="#output">Output files</a> can be used to
plot ecological and evolutionary dynamics of pests. Figure 3a shows the
pest population change both across and within seasons. Due to the
specific parameter values chosen, clear patterns reflecting pest cohorts
emerge. At the start of a season when the most crop is available, pests
eat and abundance increases. As less crop remains in a season, pest
abundance decreases. Mean food consumed varies over the course of a
season caused by the varying frequency of pests in different life
history stages (Figure 3c). In the first 50 time steps, there is some
consumption of pesticides 1 and 2 (Figure 3b), but both traits rapidly
evolve to negative values making both pesticides ineffective (Figure
3d). In this case, the trade-off in pesticide consumption is not strong
enough to maintain susceptibility to either pesticide in the population.
Figure 4 shows the location of pests on the landscape in the last time
of the simulation (left panel), and how different pesticides are
currently being applied on the landscape (right panel). Because the
number of crops and pesticides, the number and nature of traits, and the
size of pest genomes all vary depending on simulation parameters, no
plotting functions are introduced in the resevol package. Instead,
methods and code for producing plots such as those in Figures 3 and 4
are explained in Supporting Information 2.</p>
<div class="figure">
<img src="resevol_files/figure-html/overtime-1.png" alt="**Figure 3**. *Agricultural pest ecological and evolutionary dynamics over 160 time steps from an individual-based simulation using the resevol R package. Panels show (a) pesticide abundance change, (b) mean of pesticides 1 and 2 consumed per pest, (c) mean food consumed per pest, and (d) mean values of evolving traits 1 and 2 over time. Traits 1 and 2 define how much of pesticides 1 and 2, respectively, a pest will consume. White and grey vertical stripes indicate seasons of a single crop and pesticide application.*" width="768"><p class="caption">
<strong>Figure 3</strong>. <em>Agricultural pest ecological and
evolutionary dynamics over 160 time steps from an individual-based
simulation using the resevol R package. Panels show (a) pesticide
abundance change, (b) mean of pesticides 1 and 2 consumed per pest, (c)
mean food consumed per pest, and (d) mean values of evolving traits 1
and 2 over time. Traits 1 and 2 define how much of pesticides 1 and 2,
respectively, a pest will consume. White and grey vertical stripes
indicate seasons of a single crop and pesticide application.</em>
</p>
</div>
<div class="figure">
<img src="resevol_files/figure-html/landscape-1.png" alt="**Figure 4**. *Locations of pests (black) across a landscape that includes nine farms (coloured blocks) in the last time step of a simulation using the resevol R package (left panel). The right panel shows which farms apply pesticide 1 (dark grey) and 2 (light grey).*" width="614.4"><p class="caption">
<strong>Figure 4</strong>. <em>Locations of pests (black) across a
landscape that includes nine farms (coloured blocks) in the last time
step of a simulation using the resevol R package (left panel). The right
panel shows which farms apply pesticide 1 (dark grey) and 2 (light
grey).</em>
</p>
</div>
</div>
<div class="section level2">
<h2 id="conclusions">Conclusions<a class="anchor" aria-label="anchor" href="#conclusions"></a>
</h2>
<p>Insecticide resistance to pesticides is a widespread problem that
affects global food security <span class="citation">(<a href="#ref-Tabashnik2014" role="doc-biblioref">Tabashnik et al.
2014</a>; <a href="#ref-Bradshaw2016" role="doc-biblioref">Bradshaw et
al. 2016</a>; <a href="#ref-Lykogianni2021" role="doc-biblioref">Lykogianni et al. 2021</a>)</span>. Many models
have been developed to investigate the evolution of resistance under
different ecological and evolutionary conditions <span class="citation">(e.g., <a href="#ref-Curtis1985" role="doc-biblioref">Curtis 1985</a>; <a href="#ref-Mani1985" role="doc-biblioref">Mani 1985</a>; <a href="#ref-Ives2002" role="doc-biblioref">Ives and Andow 2002</a>; <a href="#ref-Liang2013" role="doc-biblioref">Liang et al. 2013</a>; <a href="#ref-Liang2018" role="doc-biblioref">Liang, Tang, and Cheke 2018</a>; <a href="#ref-Haridas2018" role="doc-biblioref">Haridas and Tenhumberg
2018</a>; <a href="#ref-South2018" role="doc-biblioref">South and
Hastings 2018</a>; <a href="#ref-Sudo2018" role="doc-biblioref">Sudo et
al. 2018</a>; <a href="#ref-Saikai2021" role="doc-biblioref">Saikai,
Hurley, and Mitchell 2021</a>)</span>. The resevol R package makes it
possible to rapidly develop and simulate myriad individual-based models
of resistance evolution. It also introduces a novel approach to
modelling complex pest genetic architecture, using an evolutionary
algorithm to generate haploid or diploid loci that map to pest traits
with pre-specified covariances (Figure 1; Supporting Information 1).
Agricultural landscapes and pest life histories are highly customisable,
allowing targetted models that can simulate specific real-world case
studies (Supporting Information 2). The breadth of possible models that
can be simulated with resevol also makes it a useful tool for developing
theory on pest management, and more even generally on the evolution and
ecology of individuals with complex traits on a changing landscape.</p>
</div>
<div class="section level2">
<h2 id="acknowledgements">Acknowledgements<a class="anchor" aria-label="anchor" href="#acknowledgements"></a>
</h2>
<p>This software was developed as part of the project for Enhancing
Diversity to Overcome Resistance Evolution (ENDORSE). The ENDORSE
project is a joint Newton funded international partnership between the
Biotechnology and Biological Sciences Research Council (BBSRC) in the UK
and the São Paulo Research Foundation (FAPESP) in Brazil under BBSRC
award reference BB/S018956/1 and FAPESP award reference 2018/21089-3.
ENDORSE is a partnership among Universidade Estadual Paulista (UNESP),
the University of Stirling (UoS), and the Centre for Agricultural and
Biosciences International (CABI).</p>
</div>
<div class="section level2">
<h2 id="contributions">Authors contributions<a class="anchor" aria-label="anchor" href="#contributions"></a>
</h2>
<p>All authors contributed to planning software development and writing.
ABD wrote and maintains package code. RM contributed code for data
visualisation.</p>
</div>
<div class="section level2">
<h2 id="availability">Availability<a class="anchor" aria-label="anchor" href="#availability"></a>
</h2>
<p>The resevol R package can be downloaded from CRAN (<a href="https://cran.r-project.org/package=resevol" class="external-link uri">https://cran.r-project.org/package=resevol</a>) or GitHub
(<a href="https://bradduthie.github.io/resevol/" class="external-link uri">https://bradduthie.github.io/resevol/</a>). The package is
open source under GNU Public License.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Bradshaw2016" class="csl-entry">
Bradshaw, Corey J A, Boris Leroy, Céline Bellard, David Roiz, Céline
Albert, Alice Fournier, Morgane Barbet-Massin, Jean Michel Salles,
Frédéric Simard, and Franck Courchamp. 2016. <span>“<span class="nocase">Massive yet grossly underestimated global costs of
invasive insects</span>.”</span> <em>Nature Communications</em> 7:
12986. <a href="https://doi.org/10.1038/ncomms12986" class="external-link">https://doi.org/10.1038/ncomms12986</a>.
</div>
<div id="ref-Busi2013" class="csl-entry">
Busi, Roberto, Paul Neve, and Stephen Powles. 2013. <span>“<span class="nocase">Evolved polygenic herbicide resistance in Lolium rigidum
by low-dose herbicide selection within standing genetic
variation</span>.”</span> <em>Evolutionary Applications</em> 6 (2):
231–42. <a href="https://doi.org/10.1111/j.1752-4571.2012.00282.x" class="external-link">https://doi.org/10.1111/j.1752-4571.2012.00282.x</a>.
</div>
<div id="ref-Carriere2016" class="csl-entry">
Carrière, Yves, Jeffrey A. Fabrick, and Bruce E. Tabashnik. 2016.
<span>“<span class="nocase">Can Pyramids and Seed Mixtures Delay
Resistance to Bt Crops?</span>”</span> <em>Trends in Biotechnology</em>
34 (4): 291–302. <a href="https://doi.org/10.1016/j.tibtech.2015.12.011" class="external-link">https://doi.org/10.1016/j.tibtech.2015.12.011</a>.
</div>
<div id="ref-Curtis1985" class="csl-entry">
Curtis, C. F. 1985. <span>“<span class="nocase">Theoretical models of
the use of insecticide mixtures for the management of
resistance</span>.”</span> <em>Bulletin of Entomological Research</em>
75 (2): 259–66. <a href="https://doi.org/10.1017/S0007485300014346" class="external-link">https://doi.org/10.1017/S0007485300014346</a>.
</div>
<div id="ref-Duthie2018" class="csl-entry">
Duthie, A Bradley, Jeremy J Cusack, Isabel L Jones, Erlend B Nilsen,
Rocio A Pozo, O. Sarobidy Rakotonarivo, Bram Van Moorter, and Nils
Bunnefeld. 2018. <span>“<span class="nocase">GMSE: an R package for
generalised management strategy evaluation</span>.”</span> <em>Methods
in Ecology and Evolution</em> 9: 2396–2401. <a href="https://doi.org/10.1101/221432" class="external-link">https://doi.org/10.1101/221432</a>.
</div>
<div id="ref-Elnahal2022" class="csl-entry">
Elnahal, Ahmed SM, Mohamed T El-Saadony, Ahmed M Saad, El Sayed M
Desoky, Amira M El-Tahan, Mostafa M Rady, Synan F AbuQamar, and Khaled A
El-Tarabily. 2022. <em><span class="nocase">The use of microbial
inoculants for biological control, plant growth promotion, and
sustainable agriculture: A review</span></em>. Vol. 162. 4. Springer
Netherlands. <a href="https://doi.org/10.1007/s10658-021-02393-7" class="external-link">https://doi.org/10.1007/s10658-021-02393-7</a>.
</div>
<div id="ref-Ffrench-Constant2013" class="csl-entry">
Ffrench-Constant, Richard H. 2013. <span>“<span class="nocase">The
molecular genetics of insecticide resistance</span>.”</span>
<em>Genetics</em> 194 (4): 807–15. <a href="https://doi.org/10.1534/genetics.112.141895" class="external-link">https://doi.org/10.1534/genetics.112.141895</a>.
</div>
<div id="ref-Ffrench-Constant2004" class="csl-entry">
Ffrench-Constant, Richard H, Phillip J Daborn, and Gaelle Le Goff. 2004.
<span>“<span class="nocase">The genetics and genomics of insecticide
resistance</span>.”</span> <em>Trends in Genetics</em> 20 (3): 163–70.
<a href="https://doi.org/10.1016/j.tig.2004.01.003" class="external-link">https://doi.org/10.1016/j.tig.2004.01.003</a>.
</div>
<div id="ref-Firko1990" class="csl-entry">
Firko, Michael J, and Jane Leslie Hayes. 1990. <span>“<span class="nocase">Quantitative genetic tools for insecticide resistance
risk assessment: estimating the heritability of
resistance</span>.”</span> <em>Journal of Economic Entomology</em> 83
(3): 657–54.
</div>
<div id="ref-Gould2018" class="csl-entry">
Gould, Fred, Zachary S. Brown, and Jennifer Kuzma. 2018. <span>“<span class="nocase">Wicked evolution: Can we address the sociobiological
dilemma of pesticide resistance?</span>”</span> <em>Science</em> 360
(6390): 728–32. <a href="https://doi.org/10.1126/science.aar3780" class="external-link">https://doi.org/10.1126/science.aar3780</a>.
</div>
<div id="ref-Green2020" class="csl-entry">
Green, Kristina Karlsson, Johan A Stenberg, and Åsa Lankinen. 2020.
<span>“<span class="nocase">Making sense of Integrated Pest Management
(IPM) in the light of evolution</span>.”</span> <em>Evolutionary
Applications</em> 13 (8): 1791–1805. <a href="https://doi.org/10.1111/eva.13067" class="external-link">https://doi.org/10.1111/eva.13067</a>.
</div>
<div id="ref-Hamblin2013" class="csl-entry">
Hamblin, Steven. 2013. <span>“<span class="nocase">On the practical
usage of genetic algorithms in ecology and evolution</span>.”</span>
<em>Methods in Ecology and Evolution</em> 4 (2): 184–94. <a href="https://doi.org/10.1111/2041-210X.12000" class="external-link">https://doi.org/10.1111/2041-210X.12000</a>.
</div>
<div id="ref-Haridas2018" class="csl-entry">
Haridas, C. V., and Brigitte Tenhumberg. 2018. <span>“<span class="nocase">Modeling effects of ecological factors on evolution of
polygenic pesticide resistance</span>.”</span> <em>Journal of
Theoretical Biology</em> 456: 224–32. <a href="https://doi.org/10.1016/j.jtbi.2018.07.034" class="external-link">https://doi.org/10.1016/j.jtbi.2018.07.034</a>.
</div>
<div id="ref-Hawkins2019" class="csl-entry">
Hawkins, Nichola J., Chris Bass, Andrea Dixon, and Paul Neve. 2019.
<span>“<span class="nocase">The evolutionary origins of pesticide
resistance</span>.”</span> <em>Biological Reviews</em> 94 (1): 135–55.
<a href="https://doi.org/10.1111/brv.12440" class="external-link">https://doi.org/10.1111/brv.12440</a>.
</div>
<div id="ref-Ives2002" class="csl-entry">
Ives, Anthony R., and D. A. Andow. 2002. <span>“<span class="nocase">Evolution of resistance to Bt crops: Directional
selection in structured environments</span>.”</span> <em>Ecology
Letters</em> 5 (6): 792–801. <a href="https://doi.org/10.1046/j.1461-0248.2002.00392.x" class="external-link">https://doi.org/10.1046/j.1461-0248.2002.00392.x</a>.
</div>
<div id="ref-Liang2018" class="csl-entry">
Liang, Juhua, Sanyi Tang, and Robert A. Cheke. 2018. <span>“<span class="nocase">A discrete host-parasitoid model with development of
pesticide resistance and ipm strategies</span>.”</span> <em>Journal of
Biological Dynamics</em> 12 (1): 1059–78. <a href="https://doi.org/10.1080/17513758.2018.1556351" class="external-link">https://doi.org/10.1080/17513758.2018.1556351</a>.
</div>
<div id="ref-Liang2013" class="csl-entry">
Liang, Juhua, Sanyi Tang, Juan J Nieto, and Robert A Cheke. 2013.
<span>“<span class="nocase">Analytical methods for detecting pesticide
switches with evolution of pesticide resistance</span>.”</span>
<em>Mathematical Biosciences</em> 245 (2): 249–57. <a href="https://doi.org/10.1016/j.mbs.2013.07.008" class="external-link">https://doi.org/10.1016/j.mbs.2013.07.008</a>.
</div>
<div id="ref-Lykogianni2021" class="csl-entry">
Lykogianni, Maira, Eleftheria Bempelou, Filitsa Karamaouna, and
Konstantinos A Aliferis. 2021. <span>“<span class="nocase">Do pesticides
promote or hinder sustainability in agriculture? The challenge of
sustainable use of pesticides in modern agriculture</span>.”</span>
<em>Science of the Total Environment</em> 795: 148625. <a href="https://doi.org/10.1016/j.scitotenv.2021.148625" class="external-link">https://doi.org/10.1016/j.scitotenv.2021.148625</a>.
</div>
<div id="ref-Mani1985" class="csl-entry">
Mani, G S. 1985. <span>“<span class="nocase">Evolution of resistance in
the presence of two insecticides.</span>”</span> <em>Genetics</em> 109
(4): 761–83. <a href="https://doi.org/10.1093/genetics/109.4.761" class="external-link">https://doi.org/10.1093/genetics/109.4.761</a>.
</div>
<div id="ref-Raymond2001" class="csl-entry">
Raymond, Michel, Claire Berticat, Mylene Weill, Nicole Pasteur, and
Christine Chevillon. 2001. <span>“<span class="nocase">Insecticide
resistance in the mosquito Culex pipiens: What have we learned about
adaptation?</span>”</span> <em>Genetica</em> 103 (3): 239–48. <a href="https://doi.org/10.1023/A" class="external-link">https://doi.org/10.1023/A</a>.
</div>
<div id="ref-Saikai2021" class="csl-entry">
Saikai, Yuji, Terrance M. Hurley, and Paul D. Mitchell. 2021.
<span>“<span class="nocase">An agent-based model of insect resistance
management and mitigation for Bt maize: a social science
perspective</span>.”</span> <em>Pest Management Science</em> 77 (1):
273–84. <a href="https://doi.org/10.1002/ps.6016" class="external-link">https://doi.org/10.1002/ps.6016</a>.
</div>
<div id="ref-South2018" class="csl-entry">
South, Andy, and Ian M Hastings. 2018. <span>“<span class="nocase">Insecticide resistance evolution with mixtures and
sequences: A model-based explanation</span>.”</span> <em>Malaria
Journal</em> 17 (1): 1–20. <a href="https://doi.org/10.1186/s12936-018-2203-y" class="external-link">https://doi.org/10.1186/s12936-018-2203-y</a>.
</div>
<div id="ref-Sudo2018" class="csl-entry">
Sudo, Masaaki, Daisuke Takahashi, David A Andow, Yoshito Suzuki, and
Takehiko Yamanaka. 2018. <span>“<span class="nocase">Optimal management
strategy of insecticide resistance under various insect life histories:
Heterogeneous timing of selection and interpatch
dispersal</span>.”</span> <em>Evolutionary Applications</em> 11 (2):
271–83. <a href="https://doi.org/10.1111/eva.12550" class="external-link">https://doi.org/10.1111/eva.12550</a>.
</div>
<div id="ref-Tabashnik2014" class="csl-entry">
Tabashnik, Bruce E, David Mota-Sanchez, Mark E Whalon, Robert M
Hollingworth, and Yves Carrière. 2014. <span>“<span class="nocase">Defining terms for proactive management of resistance to
Bt crops and pesticides</span>.”</span> <em>Journal of Economic
Entomology</em> 107 (2): 496–507. <a href="https://doi.org/10.1603/EC13458" class="external-link">https://doi.org/10.1603/EC13458</a>.
</div>
<div id="ref-Via1986" class="csl-entry">
Via, Sara. 1986. <span>“<span class="nocase">Quantitative genetic models
and the evolution of pesticide resistance</span>.”</span> In
<em>Pesticide Resistance: Strategies and Tactics for Management</em>,
222–35. Washington, DC: National Academy Press.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by A. Bradley Duthie, Rose McKeon.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
